community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArrayHashTable.java

```diff
@@ -0,0 +1,174 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime;
+
+import org.neo4j.helpers.collection.Pair;
+
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.CONTINUE_PROBING;
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.NOT_IN_USE;
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.SLOT_EMPTY;
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.VALUE_FOUND;
+
+/**
+ * This hash table is used as a backing store for the keys of set, map and multi-map where the keys are arrays of longs.
+ */
+class LongArrayHashTable
+{
+    public final long[] keys;
+    public final int width;
+    public final int capacity;
+    private int numberOfEntries;
+    private int resizeLimit;
+
+    private static final double LOAD_FACTOR = 0.75;
+
+    int tableMask;
+
+    LongArrayHashTable( int capacity, int width )
+    {
+        resizeLimit = (int) (capacity * LOAD_FACTOR);
+        tableMask = Integer.highestOneBit( capacity ) - 1;
+        keys = new long[capacity * width];
+        this.width = width;
+        java.util.Arrays.fill( keys, NOT_IN_USE );
+        this.capacity = capacity;
+    }
+
+    /**
+     * Signals whether it's time to size up or not. The actual resizing is done slightly differently depending on if this is a map or set. Maps have, in
+     * addition to a hash table also a separate array for the values.
+     *
+     * @return true if the number of keys in the hash table has reached capacity.
+     */
+    boolean timeToResize()
+    {
+        return numberOfEntries == resizeLimit;
+    }
+
+    /***
+     * Checks whether a slot in the table contains a given key.
+     * @param slot Slot to check
+     * @param key Key to check
+     * @return Can return:
+     *  SLOT_EMPTY - This slot is free. In other words - the key is not in the table
+     *  CONTINUE_PROBING - This slot is taken by a different key
+     *  VALUE_FOUND - The key is in the table at this slot
+     */
+    int checkSlot( int slot, long[] key )
+    {
+        assert LongArrayHash.validValue( key, width );
+
+        int startOffset = slot * width;
+        if ( keys[startOffset] == NOT_IN_USE )
+        {
+            return SLOT_EMPTY;
+        }
+
+        for ( int i = 0; i < width; i++ )
+        {
+            if ( keys[startOffset + i] != key[i] )
+            {
+                return CONTINUE_PROBING;
+            }
+        }
+
+        return VALUE_FOUND;
+    }
+
+    /**
+     * Writes the key to this slot.
+     *
+     * @param slot The slot to write to.
+     * @param key Le key.
+     */
+    void claimSlot( int slot, long[] key )
```

Comment:
I think the contract for this method requires that we do not overwrite slots, because otherwise `numberOfEntries` will be broken. Add assert? Allow overwrite?